### DISCUSSION TITLE
PFC8574 and/or MCP230XX on ESP32

### DISCUSSION THREAD

**Discussion by thermseekr** [arendst/Tasmota#15796] [Q&A] [answered]
Hi everyone,

Quick question, is the PFC8574 driver available for the ESP32?

Also, I checked the documentation and also #11176 and I see the expander has support for inputs. Do you think using a PFC8574 as inputs to connect momentary switches is a good idea, I mean is it as reliable as using the native GPIOs, or better to use only the native pins for this purpose and leave the expander for the outputs?

I'm planning on building a 16 in/16 out board driven by an ESP32 and I'm evaluating my options.

Thank you.

Regards,
Tales

**Comment by barbudor** [comment, marked as answer]
The PCF8574 is available on either ESP8266 or ESP32 but it isn't available in any default builds. Check : https://tasmota.github.io/docs/BUILDS/
So you will have to build your own binary. See https://tasmota.github.io/docs/Compile-your-build/

The PCF8574 or the MCP230xx can be used for inputs but through rules (there is no direct software connection between a extender's input and a relay.
The disadvantage vs native GPIO is that you won't benefit from the debouncing so you will manage to have a clean signal on the hardware side.
On the other side, the extenders outputs can be used as Relay with same features as Relay on native GPIO

So for your 16/16 device, I would choose to have the outputs on the extender and the inputs on native GPIO

**Comment by thermseekr** [comment]
Hello all,

I see there is a new Mode 2 introduced in release 12.4 that enables I/O pins on expanders to function as native GPIOs. Is it really that simple? I mean, only a JSON template describing the I/Os and that's it? This is so cool! I developed an ESP32 board with 32 inputs and 32 outputs and ethernet, and I'm wanting to confirm this is really it before sending an order to JLCPCB.

Also, I'm confused about using or not using the interrupt pins on the MCPs. Are they needed in Mode 2? And in Mode 1, what is the difference between using them or not? Doesn't the driver already poll the bus every 50ms to check for changes? So what is the use of an interrupt, make the ESP respond faster than the 50ms interval?

Thank you!

**Reply by barbudor** [reply]
Hi
Beware that not all type of GPIO are supported. Only Relay, buttons and switches but including all variations and options (inverted and not inverted, multi-press buttons, switchmode, debouncing ) 
But do not expect to map i2c, Ds18x20 or any other driver to a PCF pin. 

Interrupteur pins are not used in mode 1
In mode 2 it allows a faster reaction to inputs as switches/buttons (almost immediate VS 50ms polling)

**Reply by barbudor** [reply]
Note that for 32 input/32 ouput Tasmota is still limited to a maximum of
- 32 relays
- 28 switches
- 32 buttons
Across all GPIO (native and through port extenders)

**Reply by thermseekr** [reply]
Hi! 👋🏻 

I use these boards to control lighting in industrial and office buildings at the company I work for. So the types of GPIO supported in the expanders is enough for me as I use only switches and relays in this project.

I'm currently using an 8in 8out version with onboard relays and an Olimex ESP32 attached as a daughter board. The system works with 24VDC which is the standard in industrial automation. The idea now is to expand the IO count, migrate the ESP32 and Ethernet to the main board, and move the relays to external modules. I can live with a few buttons since Tasmota is limited to 28 switches (I was not aware of that, thanks for the heads up). Or I can start with a cheaper/smaller 16in 16out version to see how Ethernet will behave in my design.

About interrupts, for some reason I thought they were used in mode 1 and not in mode 2. I didn't find any information in the docs on how to connect them to the ESP and how/where to declare those pins. Did I go through the wiki too fast?

Thank you @barbudor for replying so fast and sharing such useful information.

Regards from Brazil 🇧🇷 
Tales

**Reply by arendst** [reply]
PCF8574 

- Mode 1 doesn't use interrupt signal (GPIO) but polls input changes every 50 mS
- Mode 2 use interrupt signal (GPIO) to set a flag which in turn is polled every loop which in turn polls input changes. If no interrupt gpio is configured it polls input changes every 100 mS.

MCP23xxx 

- Mode 1 does use interrupt signal (Register) by polling it's state every 50 mS.
- Mode 2 use interrupt signal (GPIO) to set a flag which in turn is polled every loop which in turn polls input changes. If no interrupt gpio is configured it polls input changes every 50 mS.

**Reply by thermseekr** [reply]
Hi Theo

Thanks for clearing how modes are structured. One thing tho I still haven't understood, and didn't find anything about it in the docs. I'll be using MCP23016 so I'll be going to Mode 2 with interrupts. How do I tell the driver to which GPIOs (on the ESP) I have connected the actual interrupt pins from the MCPs?

**Reply by barbudor** [reply]
You should set the GPIO to "MCP23xxx Int"

**Comment by thermseekr** [comment]
Ah, in the Web UI, now I see it. Sorry, I was thinking on how to put that info in the template and totally forgot about the Web UI. Thank you @barbudor and @arendst .

**Comment by thermseekr** [comment]
Added the following text to the docs:

`In Mode 2 you can choose to connect the interrupt pins from the MCP230xx to native GPIOs on the ESP. You will then need to configure the used GPIOs as "MCP23xxx Int" in the Web UI. This way, upon a change in an input, the corresponding interrupt pin will gererate a flag that will be scanned at every program cycle, making the detection as fast as possible. If not using the interrupt pins the detection will be triggered every 50ms.`

**Reply by thermseekr** [reply]
Prototypes are here! 16in/16out version.

![image](https://github.com/arendst/Tasmota/assets/48183960/c8ff1d06-6e5c-4989-a2cc-374587a58f69)

![image](https://github.com/arendst/Tasmota/assets/48183960/967c6652-e032-4dd7-adc7-383349c2f019)

Plugged USB as soon as I got the board out of the box, successfully transferred tasmota32.

Is there a way to set pin functions (in, out, PWM, etc) on compilation ou via serial, or does it need to be done via the web UI? I had a look but found no info about it.

Cheers
Tales

**Reply by sfromis** [reply]
You can have a template config baked into the binary, with all the relevant gpio pins. 
`#define USER_TEMPLATE "..."`
All the `"` inside the template string needs to be escaped as `\" `

To have the template activated by default:
`#define MODULE USER_MODULE` 

You can get the template string via "Configure Template" in the web UI, and then copy/paste from "Configure Other".

To configure via serial, you can use the `Template` command to set the template, and `Module 0`

**Reply by HofeBY** [reply]
Maybe own compiled TM13.0.0.2 Version in DE help you.

https://github.com/HofeBY/Tasmota-Firmware

If you need it in your own language change Location and language in config files and run throug your compiler.

**Comment by thermseekr** [comment]
Hello all,

Prototypes are working very well so far. There are still a few tests to do, like ethernet transfer speed, signals levels/integrity, and if Tasmota is correctly scanning the interrupt pins. I'll measure reaction times with and without interrupts configured and see what changes.

There was one hiccup at the very start: a protection diode was inverted on the PCB. Gotta check better the components positions before confirming the order at JLCPCB next time. And also LEDs can be less bright :)

Thank you very, very much for all the help and for making Tasmota the absolute wonder it is. What a piece of software! 🥇🥇 

![IMG_0569](https://github.com/arendst/Tasmota/assets/48183960/750ee5f7-3b14-4053-84b2-0dadc592725b)

https://github.com/arendst/Tasmota/assets/48183960/621cc468-54ff-441f-938f-016d403ee6ac
